reconcile/utils/early_exit_cache.py

```diff
@@ -64,24 +115,112 @@ def get(self, key: CacheKey) -> CacheValue:
+    def set(
+        self,
+        key: CacheKey,
+        value: CacheValue,
+        ttl_seconds: int,
+        latest_cache_source_digest: str,
+    ) -> None:
+        """
+        Set the cache value for the given key.
+
+        :param key: cache key
+        :param value: cache value
+        :param ttl_seconds: time to live in seconds
+        :param latest_cache_source_digest: latest cache source digest, used to check stale for dry run cache
+        :return: None
+        """
         expire_at = datetime.now(tz=UTC) + timedelta(seconds=ttl_seconds)
-        metadata = {EXPIRE_AT_METADATA_KEY: str(int(expire_at.timestamp()))}
+        metadata = {
+            EXPIRE_AT_METADATA_KEY: str(int(expire_at.timestamp())),
+            CACHE_SOURCE_DIGEST_METADATA_KEY: key.cache_source_digest,
+            LATEST_CACHE_SOURCE_DIGEST_METADATA_KEY: latest_cache_source_digest,
+        }
         self.state.add(
             str(key),
             value.dict(),
             metadata=metadata,
             force=True,
         )
 
-    def head(self, key: CacheKey) -> CacheStatus:
-        exists, metadata = self.state.head(str(key))
-        if not exists:
-            return CacheStatus.MISS
+    def head(self, key: CacheKey) -> CacheHeadResult:
+        """
+        Get the cache status and latest cache source digest for the given key.
+        Additionally, for dry run key, it will use latest cache source digest to check stale.
+
+        :param key: CacheKey
+        :return: CacheHeadResult
+        """
+        return self._head_dry_run(key) if key.dry_run else self._head_no_dry_run(key)
+
+    @staticmethod
+    def _is_stale(
+        metadata: Mapping[str, str],
+        latest_cache_source_digest: str,
+    ) -> bool:
+        cached_latest_cache_source_digest = (
+            metadata.get(LATEST_CACHE_SOURCE_DIGEST_METADATA_KEY) or ""
+        )
+        return cached_latest_cache_source_digest != latest_cache_source_digest
 
+    @staticmethod
+    def _is_expired(metadata: Mapping[str, str]) -> bool:
         expire_at = datetime.fromtimestamp(
             int(metadata[EXPIRE_AT_METADATA_KEY]),
             tz=UTC,
         )
         now = datetime.now(UTC)
-        return CacheStatus.HIT if now < expire_at else CacheStatus.EXPIRED
+        return now >= expire_at
+
+    def _head_dry_run_status(
+        self,
+        key: CacheKey,
+        latest_cache_source_digest: str,
+    ) -> CacheStatus:
+        exists, metadata = self.state.head(key.dry_run_path())
+        if not exists:
+            return CacheStatus.MISS
+        if self._is_expired(metadata):
+            return CacheStatus.EXPIRED
+        if self._is_stale(metadata, latest_cache_source_digest):
+            return CacheStatus.STALE
+        return CacheStatus.HIT
+
+    def _head_dry_run(self, key: CacheKey) -> CacheHeadResult:
+        _, latest_metadata = self.state.head(key.no_dry_run_path())
```

Comment:
Which shard will `key.no_dry_run_path()` use in dry-run?

Reply:
But, how does that work? In MR checks we do not ruse sharding, or are you going to change this?

Reply:
I plan to ask why it's disabled, it was running with shard before.

When it's not used, which means there is no matching latest cache in no dry run, so stale check will always pass (not stale).

When both pr check and real run are running with the same shard model, the stale check can work as expected.

Reply:
Wouldn't it be safer in this case to not do an early exit?

Reply:
Then it means the cache is not usable for pr check at all. So far the stale case is an edge case, most changes just forward changes, do revert and hit an old cache within 1 hour is rare.

Reply:
Then it will run without shard, means no matching `no-dry-run` cache, stale check is disabled. If we can override TTL (extra-args) in shard mode to be a long one, no shard mode to be a small one, the risk of showing staled output in PR check should be reduced.

Reply:
I'll merge this to test run, set pr check cache TTL to 30 mins to avoid edge cases, and can cover most rebase cases.